### ISSUE TITLE
Starting with JupyterLab 0.33.0, webpack command not found

### ISSUE THREAD

**Issue by sappjw** [jupyterlab/jupyterlab#5071]
I'm not sure if this is a `conda`/conda-forge or JupyterLab problem, but it seems most likely to me that it's with JupyterLab.  Starting with 0.33.0, I get the following error message when trying to install a labextension:

    > node ~/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build:prod
    yarn run v1.6.0
    $ webpack --config webpack.prod.config.js
    /bin/sh: webpack: command not found
    error Command failed with exit code 127.
    info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

This doesn't happen when I have 0.32.1 or below installed.  I don't have a `webpack` command on my `$PATH`.  I've also obtained the same result when running the `node` command manually from the staging directory.  The following is the test I ran in a clean `conda` environment (i.e., I only specified that `jupyterlab` should be installed) that failed to install the extension:

    (jlab_test) $ conda install jupyterlab=0.33.0
    Solving environment: done

    ## Package Plan ##

      environment location: ~/.miniconda3/envs/jlab_test

      added / updated specs:
        - jupyterlab=0.33.0

    The following packages will be downloaded:

        package                    |            build
        ---------------------------|-----------------
        jupyterlab-0.33.0          |           py36_0        10.0 MB  conda-forge

    The following packages will be UPDATED:

        jupyterlab:          0.32.0-py36_1 conda-forge --> 0.33.0-py36_0 conda-forge
        jupyterlab_launcher: 0.10.5-py36_0 conda-forge --> 0.11.2-py_0   conda-forge

    Proceed ([y]/n)? y

    Downloading and Extracting Packages
    jupyterlab-0.33.0    | 10.0 MB | #################################################################################################################################################################################################### | 100%
    Preparing transaction: done
    Verifying transaction: done
    Executing transaction: done
    (jlab_test) $ jupyter lab clean
    (jlab_test) $ jupyter labextension install @pyviz/jupyterlab_pyviz
    > ~/.miniconda3/bin/npm pack @pyviz/jupyterlab_pyviz
    pyviz-jupyterlab_pyviz-0.6.0.tgz
    > node ~/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js install
    yarn install v1.6.0
    info No lockfile found.
    [1/4] Resolving packages...
    warning css-loader > cssnano > autoprefixer > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    warning css-loader > cssnano > postcss-merge-rules > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    warning css-loader > cssnano > postcss-merge-rules > caniuse-api > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    [2/4] Fetching packages...
    info fsevents@1.2.4: The platform "linux" is incompatible with this module.
    info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
    [3/4] Linking dependencies...
    warning "@jupyterlab/vdom-extension > @nteract/transform-vdom@1.1.1" has incorrect peer dependency "react@^15.6.1".
    [4/4] Building fresh packages...
    success Saved lockfile.
    Done in 108.20s.
    > node ~/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build:prod
    yarn run v1.6.0
    $ webpack --config webpack.prod.config.js
    /bin/sh: webpack: command not found
    error Command failed with exit code 127.
    info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

This is the result with version 0.32.1 (the one prior to 0.33.0 that is currently available in conda-forge):

    (jlab_test) $ conda install jupyterlab=0.32.1
    Solving environment: done

    ## Package Plan ##

      environment location: ~/.miniconda3/envs/jlab_test

      added / updated specs:
        - jupyterlab=0.32.1

    The following packages will be DOWNGRADED:

        jupyterlab:          0.33.0-py36_0 conda-forge --> 0.32.1-py36_0 conda-forge
        jupyterlab_launcher: 0.11.2-py_0   conda-forge --> 0.10.5-py36_0 conda-forge

    Proceed ([y]/n)? y

    Preparing transaction: done
    Verifying transaction: done
    Executing transaction: done
    (jlab_test) $ jupyter labextension install @pyviz/jupyterlab_pyviz
    > ~/.miniconda3/bin/npm pack @pyviz/jupyterlab_pyviz
    pyviz-jupyterlab_pyviz-0.6.0.tgz
    > ~/.miniconda3/bin/npm pack @pyviz/jupyterlab_pyviz@0.5.3
    pyviz-jupyterlab_pyviz-0.5.3.tgz
    Incompatible extension:

    "@pyviz/jupyterlab_pyviz@0.6.0" is not compatible with the current JupyterLab
    Conflicting Dependencies:
    JupyterLab              Extension        Package
    >=0.16.3 <0.17.0        >=0.17.0 <0.18.0 @jupyterlab/application
    >=0.16.4 <0.17.0        >=0.17.0 <0.18.0 @jupyterlab/apputils
    >=0.16.3 <0.17.0        >=0.17.0 <0.18.0 @jupyterlab/notebook
    >=2.0.3 <3.0.0          >=3.0.0 <4.0.0   @jupyterlab/services

    Found compatible version: 0.5.3
    > ~/.miniconda3/bin/npm pack @pyviz/jupyterlab_pyviz@0.5.3
    pyviz-jupyterlab_pyviz-0.5.3.tgz
    > node ~/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js install
    yarn install v1.5.1
    info No lockfile found.
    [1/4] Resolving packages...
    warning css-loader > cssnano > autoprefixer > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    warning css-loader > cssnano > postcss-merge-rules > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    warning css-loader > cssnano > postcss-merge-rules > caniuse-api > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    [2/4] Fetching packages...
    info fsevents@1.2.4: The platform "linux" is incompatible with this module.
    info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
    [3/4] Linking dependencies...
    warning "@jupyterlab/json-extension > react-json-tree@0.10.9" has incorrect peer dependency "react@^15.0.0".
    warning "@jupyterlab/json-extension > react-json-tree@0.10.9" has incorrect peer dependency "react-dom@^15.0.0".
    warning "@jupyterlab/vdom-extension > @nteract/transform-vdom@1.1.1" has incorrect peer dependency "react@^15.6.1".
    [4/4] Building fresh packages...
    success Saved lockfile.
    Done in 105.35s.
    > node ~/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build:prod
    yarn run v1.5.1
    $ webpack --config webpack.prod.config.js
    Hash: 93e01eaadcb105650b22
    Version: webpack 2.7.0
    Time: 49735ms
                                     Asset     Size  Chunks                    Chunk Names
            vendor.93fdee4f2b45e28305c1.js  1.57 MB       2  [emitted]  [big]  vendor
      912ec66d7572ff821749319396470bde.svg   444 kB          [emitted]  [big]
      674f50d287a8c48dc19ba404d20fe713.eot   166 kB          [emitted]
     fee66e712a8a08eef5805a46892932ad.woff    98 kB          [emitted]
    af7ae505a9eed503f8b8e6982036873e.woff2  77.2 kB          [emitted]
                 0.279ec850a9242db04c85.js   465 kB       0  [emitted]  [big]
              main.e72a74a38136522a2ade.js  2.37 MB       1  [emitted]  [big]  main
      b06871f281fee6b241d60582ae9369b9.ttf   166 kB          [emitted]
          manifest.1a16b9896eadfaf8fa1c.js  1.62 kB       3  [emitted]         manifest
             0.279ec850a9242db04c85.js.map  1.35 MB       0  [emitted]
          main.e72a74a38136522a2ade.js.map  8.24 MB       1  [emitted]         main
        vendor.93fdee4f2b45e28305c1.js.map  5.69 MB       2  [emitted]         vendor
      manifest.1a16b9896eadfaf8fa1c.js.map  8.24 kB       3  [emitted]         manifest
                                index.html  1.53 kB          [emitted]
    [/sZe] ./~/@phosphor/commands/lib/index.js 33.3 kB {2} [built]
    [2Kj9] ./~/@phosphor/widgets/lib/index.js 1.38 kB {2} [built]
    [kzCB] ./~/path-posix/index.js 6.94 kB {2} [built]
       [8] multi @phosphor/algorithm @phosphor/application @phosphor/commands @phosphor/coreutils @phosphor/datagrid @phosphor/disposable @phosphor/domutils @phosphor/dragdrop @phosphor/messaging @phosphor/properties @phosphor/signaling @phosphor/virtualdom @phosphor/widgets ajv ansi_up codemirror comment-json es6-promise marked moment path-posix react react-dom sanitize-html url-parse xterm 328 bytes {2} [built]
    [q1RZ] ./~/@phosphor/datagrid/lib/index.js 793 bytes {2} [built]
       [9] multi whatwg-fetch ./build/index.out.js 40 bytes {1} [built]
    [l9te] ./~/@phosphor/application/lib/index.js 18.1 kB {2} [built]
    [lMhA] ./~/@phosphor/dragdrop/lib/index.js 33.7 kB {2} [built]
    [nh00] ./~/@phosphor/disposable/lib/index.js 4.24 kB {2} [built]
    [o73R] ./~/sanitize-html/index.js 11.2 kB {2} [built]
    [qzHn] ./~/@phosphor/coreutils/lib/index.js 721 bytes {2} [built]
    [rplX] ./~/whatwg-fetch/fetch.js 13 kB {1} [built]
    [ug+F] ./~/xterm/lib/xterm.js 52.7 kB {2} [built]
    [wr2+] ./~/@phosphor/properties/lib/index.js 6.58 kB {2} [built]
    [wuIn] ./~/comment-json/index.js 103 bytes {2} [built]
        + 2221 hidden modules
    Child html-webpack-plugin for "index.html":
        [3IRH] (webpack)/buildin/module.js 517 bytes {0} [built]
        [DuR2] (webpack)/buildin/global.js 509 bytes {0} [built]
        [GTAU] ./~/html-loader!./templates/partial.html 420 bytes {0} [built]
        [M4fF] ./~/lodash/lodash.js 540 kB {0} [built]
        [vxCX] ./~/html-webpack-plugin/lib/loader.js!./templates/template.html 1.41 kB {0} [built]
    Done in 51.88s.

**Comment by blink1073** [comment]
Hi @sappjw, what OS are you using?

**Comment by sappjw** [comment]
CentOS Linux 6.10

**Comment by blink1073** [comment]
I was just able to install the `pyviz` extension on CentOS 6.9, can you please try removing the application directory (found with `jupyter lab paths`) and then try installing the extension again?

**Comment by sappjw** [comment]
Tried that and still: `/bin/sh: webpack: command not found` seems to be the failing command.
I forgot to note this detail before, but after I try installing the extension (which behaves the same as above) I listed the installed extensions

    $ jupyter labextension list
    JupyterLab v0.33.0
    Known labextensions:
       app dir: /home/jsapp/.miniconda3/envs/jlab_test/share/jupyter/lab
    @pyviz/jupyterlab_pyviz
            @pyviz/jupyterlab_pyviz v0.6.0  enabled  OK
    
    Build recommended, please run `jupyter lab build`:
        @pyviz/jupyterlab_pyviz needs to be included in build

Then doing what is suggested:

    $ jupyter lab build
    [LabBuildApp] > node /home/jsapp/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js install
    yarn install v1.6.0
    info No lockfile found.
    [1/4] Resolving packages...
    warning css-loader > cssnano > autoprefixer > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    warning css-loader > cssnano > postcss-merge-rules > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    warning css-loader > cssnano > postcss-merge-rules > caniuse-api > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
    [2/4] Fetching packages...
    info fsevents@1.2.4: The platform "linux" is incompatible with this module.
    info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
    [3/4] Linking dependencies...
    warning "@jupyterlab/vdom-extension > @nteract/transform-vdom@1.1.1" has incorrect peer dependency "react@^15.6.1".
    [4/4] Building fresh packages...
    success Saved lockfile.
    Done in 31.53s.
    [LabBuildApp] > node /home/jsapp/.miniconda3/envs/jlab_test/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build:prod
    yarn run v1.6.0
    $ webpack --config webpack.prod.config.js
    /bin/sh: webpack: command not found
    error Command failed with exit code 127.
    info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

**Comment by blink1073** [comment]
Can you please do the following?

- Run `jupyter lab paths`
- `cd` to the Application Directory
- `cd staging`
- `node_modules/.bin/webpack-cli --version`

**Comment by sappjw** [comment]
3.1.0

**Comment by sappjw** [comment]
This is with `jupyterlab=0.33.0` installed:

    (jlab_test) [jsapp ~/.miniconda3/envs/jlab_test/share/jupyter/lab/staging]$ npm run build:prod

    > @jupyterlab/application-top@0.17.2 build:prod /home/jsapp/.miniconda3/envs/jlab_test/share/jupyter/lab/staging
    > webpack --config webpack.prod.config.js

    sh: webpack: command not found
    npm ERR! file sh
    npm ERR! code ELIFECYCLE
    npm ERR! errno ENOENT
    npm ERR! syscall spawn
    npm ERR! @jupyterlab/application-top@0.17.2 build:prod: `webpack --config webpack.prod.config.js`
    npm ERR! spawn ENOENT
    npm ERR!
    npm ERR! Failed at the @jupyterlab/application-top@0.17.2 build:prod script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.


       ╭─────────────────────────────────────╮
       │                                     │
       │   Update available 5.6.0 → 6.3.0    │
       │     Run npm i -g npm to update      │
       │                                     │
       ╰─────────────────────────────────────╯


    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/jsapp/.npm/_logs/2018-08-10T12_39_10_893Z-debug.log

and

    (jlab_test) [jsapp ~/.miniconda3/envs/jlab_test/share/jupyter/lab/staging]$ jlpm run build:prod
    yarn run v1.6.0
    $ webpack --config webpack.prod.config.js
    /bin/sh: webpack: command not found
    error Command failed with exit code 127.
    info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

The log referenced says:

    0 info it worked if it ends with ok
    1 verbose cli [ '/home/jsapp/.miniconda3/bin/node',
    1 verbose cli   '/home/jsapp/.miniconda3/bin/npm',
    1 verbose cli   'run',
    1 verbose cli   'build:prod' ]
    2 info using npm@5.6.0
    3 info using node@v9.11.1
    4 verbose run-script [ 'prebuild:prod', 'build:prod', 'postbuild:prod' ]
    5 info lifecycle @jupyterlab/application-top@0.17.2~prebuild:prod: @jupyterlab/application-top@0.17.2
    6 info lifecycle @jupyterlab/application-top@0.17.2~build:prod: @jupyterlab/application-top@0.17.2
    7 verbose lifecycle @jupyterlab/application-top@0.17.2~build:prod: unsafe-perm in lifecycle true
    8 verbose lifecycle @jupyterlab/application-top@0.17.2~build:prod: PATH: /home/jsapp/.miniconda3/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/jsapp/.miniconda3/envs/jlab_test/share/jupyter/lab/staging/node_modules/.bin:/home/jsapp/.miniconda3/envs/jlab_test/bin:/home/jsapp/.miniconda3/bin:/home/jsapp/.local/Qt/5.6/bin:/home/jsapp/bin:/home/jsapp/.local/bin:/home/jsapp/bin:/home/jsapp/.local/bin:/home/jsapp/bin:/home/jsapp/.local/bin:/home/jsapp/bin:/home/jsapp/.local/bin:/usr/lib64/qt-3.3/bin:/home/jsapp/bin:/home/jsapp/.local/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin
    9 verbose lifecycle @jupyterlab/application-top@0.17.2~build:prod: CWD: /home/jsapp/.miniconda3/envs/jlab_test/share/jupyter/lab/staging
    10 silly lifecycle @jupyterlab/application-top@0.17.2~build:prod: Args: [ '-c', 'webpack --config webpack.prod.config.js' ]
    11 info lifecycle @jupyterlab/application-top@0.17.2~build:prod: Failed to exec build:prod script
    12 verbose stack Error: @jupyterlab/application-top@0.17.2 build:prod: `webpack --config webpack.prod.config.js`
    12 verbose stack spawn ENOENT
    12 verbose stack     at ChildProcess.<anonymous> (/home/jsapp/.miniconda3/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:48:18)
    12 verbose stack     at ChildProcess.emit (events.js:180:13)
    12 verbose stack     at maybeClose (internal/child_process.js:936:16)
    12 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:220:5)
    13 verbose pkgid @jupyterlab/application-top@0.17.2
    14 verbose cwd /home/jsapp/.miniconda3/envs/jlab_test/share/jupyter/lab/staging
    15 verbose Linux 2.6.32-754.2.1.el6.x86_64
    16 verbose argv "/home/jsapp/.miniconda3/bin/node" "/home/jsapp/.miniconda3/bin/npm" "run" "build:prod"
    17 verbose node v9.11.1
    18 verbose npm  v5.6.0
    19 error file sh
    20 error code ELIFECYCLE
    21 error errno ENOENT
    22 error syscall spawn
    23 error @jupyterlab/application-top@0.17.2 build:prod: `webpack --config webpack.prod.config.js`
    23 error spawn ENOENT
    24 error Failed at the @jupyterlab/application-top@0.17.2 build:prod script.
    24 error This is probably not a problem with npm. There is likely additional logging output above.
    25 verbose exit [ 1, true ]

**Comment by blink1073** [comment]
That looks like a problem with `npm` itself.  Are you able to run the webpack example app?  https://github.com/webpack/example-app/

**Comment by sappjw** [comment]
I had to install `1.3.x` for `webpack` and `webpack-dev-server`.  Also had to install `jade`, `coffee-script`, and `less` separately.  Finally, it appears `npm install` succeeded.  But I'm not able to run `jam install`:

    (base) [jsapp ~/tmp/example-app]$ jam install
    Building version tree...
    repositories checking "jquery"
    Error: Status code: 301
    at Function.exports.statusCodeError (/home/jsapp/.miniconda3/lib/node_modules/jamjs    /lib/couchdb.js:281:12)
    at ClientRequest.<anonymous> (/home/jsapp/.miniconda3/lib/node_modules/jamjs    /lib/couchdb.js:222:32)
        at ClientRequest.emit (events.js:182:13)
        at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:546:21)
        at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
        at Socket.socketOnData (_http_client.js:432:20)
        at Socket.emit (events.js:182:13)
        at addChunk (_stream_readable.js:283:12)
        at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    Failed

**Comment by sappjw** [comment]
After removing the User Settings directory as mentioned in #5157, this issue may be fixed.  I'm trying to verify and I'll close it when I can.

**Comment by llinfeng** [comment]
##TL;DR: ==>check if you have a live `jupyterlab` instance running or not?

I have stumbled into this post after observing exactly the same error message as reported in the original post. These error messages were generated when I tried to do the following:

1. Install `jupyterlab-toc` via command: `jupyter labextension install @jupyterlab/toc`;
2. Locally build `jupyterlab-vim` via command: `jlpm install; jlpm run build; jupyter labextension link .` 
```
....
 webpack: not found
error Command failed with exit code 127.
....
```

It confused me for a good 15 minutes, then realize that I am running JupyterLab on a Linux subsystem on Windows (WSL) ==> before hammering the `restart` button, I tried to close the running instance of the Jupyter-Lab server. Like all other Windows magics, this helped.

* Simple solution: kill all running instances of the JupyterLab server. I am guessing that `webpack` does not like to deal with a live kernel/instance of JupyterLab.
